File: dlearning/src/main/java/com/spring/dlearning/service/CommentService.java
package com.spring.dlearning.service;

import com.spring.dlearning.constant.PredefinedRole;
import com.spring.dlearning.dto.request.CommentRequest;
import com.spring.dlearning.dto.request.UpdateCommentRequest;
import com.spring.dlearning.dto.response.CommentResponse;
import com.spring.dlearning.dto.response.PageResponse;
import com.spring.dlearning.dto.response.UpdateCommentResponse;
import com.spring.dlearning.entity.Comment;
import com.spring.dlearning.entity.Post;
import com.spring.dlearning.entity.User;
import com.spring.dlearning.exception.AppException;
import com.spring.dlearning.exception.ErrorCode;
import com.spring.dlearning.mapper.CommentMapper;
import com.spring.dlearning.repository.CommentRepository;
import com.spring.dlearning.repository.PostRepository;
import com.spring.dlearning.repository.UserRepository;
import com.spring.dlearning.utils.SecurityUtils;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Objects;

@Service
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
@Slf4j
public class CommentService {

    CommentRepository commentRepository;
    PostRepository postRepository;
    UserRepository userRepository;
    CommentMapper commentMapper;
    BannedWordsService bannedWordsService;

    @PreAuthorize("isAuthenticated()")
    public PageResponse<CommentResponse> getCommentByPostId(Long postId, int page, int size) {
        Sort sort = Sort.by(Sort.Direction.DESC, "createdAt");
        Pageable pageable = PageRequest.of(page - 1, size, sort);

        Page<Comment> parentComments = commentRepository.findCommentByPostIdAndParentCommentIsNull(postId, pageable);

        List<CommentResponse> responses = parentComments.getContent()
                .stream()
                .map(comment -> {
                    CommentResponse response = commentMapper.toCommentResponse(comment);

                    List<CommentResponse> replies = comment.getReplies().stream()
                            .map(commentMapper::toCommentResponse)
                            .toList();
                    response.setReplies(replies);
                    return response;
                })
                .toList();

        return PageResponse.<CommentResponse>builder()
                .currentPage(page)
                .pageSize(pageable.getPageSize())
                .totalElements(parentComments.getTotalElements())
                .totalPages(parentComments.getTotalPages())
                .data(responses)
                .build();
    }

    @PreAuthorize("isAuthenticated()")
    public CommentResponse addComment(CommentRequest request) {

        String email = SecurityUtils.getCurrentUserLogin()
                .orElseThrow(() -> new AppException(ErrorCode.EMAIL_INVALID));

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_EXISTED));

        Post post = postRepository.findById(request.getPostId())
                .orElseThrow(() -> new AppException(ErrorCode.POST_ID_INVALID));

        Comment parentComment = null;
        if(request.getParentCommentId() != null) {
            parentComment = commentRepository.findById(request.getParentCommentId())
                    .orElseThrow(() -> new AppException(ErrorCode.PARENT_COMMENT_NOT_EXISTED));
        }

        if ((request.getContent() == null || request.getContent().isEmpty())) {
            throw new AppException(ErrorCode.CONTENT_COMMENT_INVALID);
        }

        if (bannedWordsService.containsBannedWords(request.getContent())) {
            throw new AppException(ErrorCode.INVALID_COMMENT_CONTENT);
        }

        Comment comment = commentMapper.toComment(request);
        comment.setUser(user);
        comment.setPost(post);
        comment.setParentComment(parentComment);

        commentRepository.save(comment);

        return commentMapper.toCommentResponse(comment);
    }

    @PreAuthorize("isAuthenticated()")
    public void deleteComment(Long commentId) {

        String email = SecurityUtils.getCurrentUserLogin()
                .orElseThrow(() -> new AppException(ErrorCode.EMAIL_INVALID));

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_EXISTED));

        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new AppException(ErrorCode.COMMENT_NOT_EXISTED));

        if (Objects.equals(user.getId(), comment.getUser().getId()) ||
                PredefinedRole.ADMIN_ROLE.equals(user.getRole().getName())) {
            commentRepository.delete(comment);
            return;
        }

        throw new AppException(ErrorCode.DELETE_COMMENT_INVALID);
    }

    @PreAuthorize("isAuthenticated()")
    public UpdateCommentResponse updateComment (Long commentId, UpdateCommentRequest request){

        if ( commentId <= 0 ){
            throw new AppException(ErrorCode.INVALID_PATH_VARIABLE_ID);
        }

        String email = SecurityUtils.getCurrentUserLogin()
                .orElseThrow(() -> new AppException(ErrorCode.EMAIL_INVALID));

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_EXISTED));

        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new AppException(ErrorCode.COMMENT_NOT_EXISTED));

        if (!Objects.equals(user.getId(), comment.getUser().getId())) {
            throw new AppException(ErrorCode.UPDATE_COMMENT_INVALID);
        }

        commentMapper.updateComment(request, comment);
        commentRepository.save(comment);

        return UpdateCommentResponse.builder()
                .id(comment.getId())
                .content(comment.getContent())
                .build();
    }

}

File: web-app/react-course/course-app/src/service/CommentService.js
import { toast } from "react-toastify";
import axios from "../utils/CustomizeAxios";

export const getCommentByPostId = async (postId, currentPage) => {
    try {
        const response = await axios.get(`api/v1/post-comment/${postId}`, {
            params: {
                page: currentPage
            }
        });
        return response.data;
    } catch (error) {
        console.error('Error fetching comments from API:', error.response || error.message);
        throw new Error(`Failed to fetch comments: ${error.response?.status || error.message}`);
    }
};

export const addComment = async (commentData) => {
    try {
        const response = await axios.post(`api/v1/add-comment`, commentData);
        console.log(response);
        if (response.data.result) {
            toast.success('Comment added successfully');
            return response.data.result;
        } else {
            toast.error(response.data.message);
            throw new Error(response.data.message);
        }
    } catch (error) {
        console.error('Error in addComment:', error);
        throw error;
    }
}

export const replyComment = async (replyData) => {
    try {
        const response = await axios.post(`api/v1/add-comment`, replyData)
        if (response.data.result) {
            toast.success('Comment added successfully')
            return response.data.result;
        } else {
            toast.error(response.data.message)
            throw new Error(response.data.message);
        }
    } catch (error) {
        console.log(error)
        throw error;
    }
}

export const deleteComment = async (commentId) => {
    try {
        const response = await axios.delete(`api/v1/delete-comment/${commentId}`);
        return response.data;
    } catch (error) {
        console.error(error);
        throw error;
    }
};

export const updateComment = async (commentId, updateContent) => {
    try {
        const response = await axios.put(`api/v1/update-comment/${commentId}`, { content: updateContent })
        if (response.data.result) {
            toast.success('Update Comment Successfully');
            return response.data.result;
        } else {
            toast.error(response.data.message);
            throw new Error(response.data.message);
        }
    } catch (error) {
        console.error('Error in service:', error);
        throw error;
    }
}